src/generic_event_type.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+#ifndef CATA_SRC_GENERIC_EVENT_TYPE_H
+#define CATA_SRC_GENERIC_EVENT_TYPE_H
+
+#include <string>
+#include <climits>
+
+#include "bodypart.h"
+#include "field.h"
+#include "translations.h"
+#include "type_id.h"
+
+template<typename T>
+class generic_factory;
+
+const generic_event_type_id NEXT_WEATHER_GENERIC_EVENT( "next_weather" );
+
+struct generic_event_type_field {
+    field_type_str_id type;
+    int intensity;
+    time_duration age;
+    int radius;
+    bool outdoor_only;
+};
+
+struct spawn_type {
+    mtype_id target;
+    int target_range;
+    int hallucination_count;
+    int real_count;
+    int min_radius;
+    int max_radius;
+};
+
+struct effect_info {
+    time_duration length = 0_seconds;
+    int intensity = 0;
+    efftype_id id;
+    bodypart_str_id target_part;
+};
+
+struct generic_event_type {
+    public:
+        friend class generic_factory<generic_event_type>;
+        bool was_loaded = false;
+        generic_event_type_id id;
+        int one_in_chance;
+        translation message;
```

Comment:
Rather than having a big list of all the things this might do, and having to iterate over all of them in `do_event`, I think a better design is to have a `std::vector<value_ptr<foo>>` where `foo` is an abstract base class with derived classes for each of the things that can happen in `do_event`.  Then e.g. you don't need a `std::vector<trait_id> traits_to_add;`, you can have an `add_trait` implementation and have multiple of those as needed.